many: re-re-enable the prompting notice backend

olivercalder opened the pull request:
Relying on the new `prompting-client` integration tests added in #16289, re-enable the prompting notice backend after fixing whatever is causing problems with those integration tests. The potential regression was not caught by existing prompting spread tests.

The real root cause appears to be that the prompting-client integration tests would hang after the subprocess which triggered the request exited, due to a problem with the output not sending depending on the timing and the tokio scheduler. And enabling the prompting notice backend changed the timing enough to fairly reliably trigger the tests hanging on at least a few test cases per run.

The issue with the test reliability was solved here: https://github.com/canonical/prompting-client/pull/304

This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-36018

A reproducer and analysis of the timing and the root cause can also be found there. #16879 is can be used to reproduce the issue and easily enable/disable the prompting notice backend.

We may not want to include all the extra debug logging in the final PR. I've split out the changes so it should be easy to drop/squash the parts we want to before merging. I would like to merge spread test improvements and the prompting notice backend enablement as separate PRs, at least.

unskip:

- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-flag-restart
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-integration-tests
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-prompt-restoration
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-smoke
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-snapd-startup
- openstack:ubuntu-26.04-64:tests/main/interfaces-requests-activates-handlers
- openstack:ubuntu-26.04-64:tests/main/snap-interfaces-requests-control

olivercalder committed:
Reapply "many: re-enable prompting notice backend (#16194)" (revert #16287)

This reverts commit d870ff4f721ed498282d35306633b9acdc084fac.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

o/i/apparmorprompting: handle edge case when creating notice backend

Handle the edge case in which there is a notice which was previously
recorded in snapd state re-recorded in the new notice backend while the
rule or prompt DB is being initialized, prior to the notice backend
being registered as a notice backend. In this case, we want to preserve
the "fresher" data, that is the data of the notice created in the
prompting notice backend, rather than overwrite it with the notice data
drained from the state.

This is truly rare, and can only occur during the the already rare edge
case that there are prompts or rules with active notices at the time
when snapd first restarts after refreshing for the first time to a
version which supports the prompting notice backends. Given that this is
the case, this particular edge case only occurs if a rule expired before
the rules backend was loaded during startup, thus causing a notice that
the rule expired. Now we'll preserve that `{"removed": "expired"}` data.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

zyga commented on tests/nightly/prompting-client-integration-tests/task.yaml:
```diff
@@ -68,6 +68,12 @@ prepare: |
 
+    echo "Enable timeouts for integration tests"
+    git fetch
+    git config user.email 'nonexistent@ubuntu.com'
```
This could be anything@example.org which is designed for ... well, examples like that.

zyga reviewed (commented):
Looks good, some nits. Ping me for non-draft review please.

olivercalder committed:
tests: improve robustness of apparmor-prompting-snapd-startup

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
many: add debug logging for prompting notices

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

o/i/apparmorprompting: remove debug logging from prompting notice backend internals

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

i/prompting: reduce and improve notices and logging

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

i/p/requestprompts: adjust debug logging

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

pedronis reviewed (commented):
@olivercalder yes, you should do a pass and reevaluate all the logging you added, and see what it makes sense to add and what doesn't, you might even want to separate finally landing this from adding logging

olivercalder commented on interfaces/prompting/requestrules/requestrules.go:
```diff
@@ -749,21 +750,26 @@ func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string, permis
+			logger.Debugf("rule was expired when new rule %q was added: %q", ruleID, maybeExpired.ID)
 			rdb.notifyRule(maybeExpired.User, maybeExpired.ID, expiredData)
+		} else {
+			logger.Debugf("WARNING: error occurred when trying to remove expired rule %q: %v", maybeExpired.ID, err)
```
Since these have a `String()` method defined, shouldn't it default to calling that method?

olivercalder commented on interfaces/prompting/requestrules/requestrules.go:
```diff
@@ -1291,6 +1298,7 @@ func (rdb *RuleDB) RemoveRule(user uint32, id prompting.IDType) (*Rule, error) {
 	// rule was affected. We want the rule fully removed, so this is fine.
 
 	data := map[string]string{"removed": "removed"}
+	logger.Debugf("rule was removed: %q", id)
```
Same here

olivercalder commented on interfaces/prompting/requestrules/requestrules.go:
```diff
@@ -1341,6 +1351,7 @@ func (rdb *RuleDB) removeRulesInternal(user uint32, rules []*Rule) error {
 	}
 
 	// Save successful, now remove rules' variants from tree
+	logger.Debugf("removed rules: %q", removedRuleIDs)
```
On what go version? At least on Go 1.25 and 1.26, `fmt.Sprintf("%q", []string{"foo", "bar"}` results in `["foo" "bar"]`, which is exactly what I want.

olivercalder commented on interfaces/prompting/requestprompts/requestprompts.go:
```diff
@@ -453,10 +456,17 @@ func (udb *userPromptDB) timeoutCallback(pdb *PromptDB, user uint32) {
 		p.sendReply(prompting.OutcomeDeny) // ignore any error, should not occur
 	}
+	if len(expiredPromptIDs) > 0 {
+		logger.Debugf("prompts expired: %q", expiredPromptIDs)
```
Again, with what go version? These logs should contain `["0000000000000001" "0000000000000002" "0000000000000003"]`, at least on recent versions

pedronis commented on interfaces/prompting/requestprompts/requestprompts.go:
```diff
@@ -411,6 +411,7 @@ func (udb *userPromptDB) remove(id prompting.IDType) (*Prompt, error) {
 // called directly outside of a `time.AfterFunc` call which initializes the
 // timer for the user prompt DB when it is first created.
 func (udb *userPromptDB) timeoutCallback(pdb *PromptDB, user uint32) {
+	logger.Debug("prompt timeout callback triggered")
```
should this log contain any kind of id or other reference to the originating prompt(s) ?

olivercalder replied:
The timeout callback fires if a user does not receive prompts API activity for some long time (10 minutes), then times out all existing prompts for the user.

This particular debug log just states that the timeout callback fired, but later in the function there are more specific logs. And at the end if there were in fact expired prompts due to the callback we log them:

```go
    logger.Debugf("prompts expired: %q", expiredPromptIDs)
```

olivercalder replied:
Try to remove the log on entry point to this function, instead ensure all exit points have a common prefix.

pedronis reviewed (commented):
question

olivercalder reviewed (commented):
Samuele suggested to make sure we don't have redundant logs on entry and exit points of functions.

andrewphelpsj commented on overlord/ifacestate/apparmorprompting/prompting.go:
```diff
@@ -135,7 +118,7 @@ func New(s *state.State) (m *InterfacesRequestsManager, retErr error) {
 	}()
 
-	rulesBackend, err := requestrules.New(notifyRule)
+	rulesBackend, err := requestrules.New(noticeBackends.ruleBackend.addNotice)
```
`requestrules.New` mentions that the passed in `notifyRule` can be called before `requestrules.New` returns.

Given that, is it possible that `noticeBackends.ruleBackend.addNotice` gets called with some new data before `bknd.addNotice` (within `noticeBackends.registerWithManager`) is called with a drained notice?

Then, we'd end up overwriting that new input with old data from the state?

pedronis reviewed (approved):
generally +1, please find a couple more reviewers though

andrewphelpsj reviewed (approved):
Thanks for the fix!
